naga/src/front/spv/mod.rs

```diff
@@ -4221,6 +4216,102 @@ impl<I: Iterator<Item = u32>> Frontend<I> {
+                    // scalar, and we'll need the type itself to get a handle to the atomic
+                    // result struct.
+                    let crate::TypeInner::Scalar(scalar) = ctx.module.types[p_base_ty_h].inner
+                    else {
+                        return Err(
+                            crate::front::atomic_upgrade::Error::CompareExchangeNonScalarBaseType
+                                .into(),
+                        );
+                    };
+
+                    // Get a handle to the atomic result struct type.
+                    let atomic_result_struct_ty_h = ctx.module.generate_predeclared_type(
+                        crate::PredeclaredType::AtomicCompareExchangeWeakResult(scalar),
+                    );
+
+                    block.extend(emitter.finish(ctx.expressions));
+
+                    // Create an expression for our atomic result
+                    let atomic_lexp_handle = {
+                        let expr = crate::Expression::AtomicResult {
+                            ty: atomic_result_struct_ty_h,
+                            comparison: true,
+                        };
+                        ctx.expressions.append(expr, span)
+                    };
+
+                    // Create an dot accessor to extract the value from the
+                    // result struct __atomic_compare_exchange_result<T> and use that
+                    // as the expression for the result_id
+                    {
+                        let expr = crate::Expression::AccessIndex {
+                            base: atomic_lexp_handle,
+                            index: 0,
+                        };
+                        let handle = ctx.expressions.append(expr, span);
+                        // Use this dot accessor as the result id's expression
+                        let _ = self.lookup_expression.insert(
+                            result_id,
+                            LookupExpression {
+                                handle,
+                                type_id: result_type_id,
+                                block_id,
+                            },
+                        );
+                    }
+
+                    emitter.start(ctx.expressions);
+
+                    // Create a statement for the op itself
+                    let stmt = crate::Statement::Atomic {
+                        pointer: p_exp_h,
+                        fun: crate::AtomicFunction::Exchange {
+                            compare: Some(c_lexp_handle),
+                        },
+                        value: v_lexp_handle,
+                        result: Some(atomic_lexp_handle),
+                    };
+                    block.push(stmt, span);
+
+                    // Store any associated global variables so we can upgrade their types later
+                    self.upgrade_atomics
+                        .extend(ctx.get_contained_global_variable(p_exp_h));
```

Comment:
(carried forward) Like the other insertions into `upgrade_atomics`, this should use `insert`, not `extend`, and apply the `?` operator to the return value of `get_contained_global_variable`, so that if `p_exp_h` cannot be traced back to some global variable, we get an error, rather than simply failing to add the variable to `upgrade_atomics`.